contracts/JBToken721.sol

```diff
@@ -0,0 +1,305 @@
+    contractUri = _contractUri;
+  }
+
+  //*********************************************************************//
+  // -------------------------- constructor ---------------------------- //
+  //*********************************************************************//
+
+  /**
+    @notice
+    Next token id to be minted.
+  */
+  uint256 private _nextTokenId;
+
+  /**
+    @notice
+    Current supply.
+  */
+  uint256 private _supply;
+
+  /**
+    @notice
+    Token base uri.
+  */
+  string private _baseUri;
+
+  /**
+    @notice
+    Custom token uri resolver, superceeds base uri.
+  */
+  IJBToken721UriResolver private _tokenUriResolver;
+
+  /**
+    @notice
+    Contract opensea-style metadata uri.
+  */
+  string private _contractUri;
+
+  /**
+    @param _name The name of the token.
+    @param _symbol The symbol that the token should be represented by.
+    @param _uri Token base URI.
+    @param _tokenUriResolverAddress Custom uri resolver.
+    @param _contractMetadataUri Contract metadata uri.
+  */
+  constructor(
+    string memory _name,
+    string memory _symbol,
+    string memory _uri,
+    IJBToken721UriResolver _tokenUriResolverAddress,
+    string memory _contractMetadataUri
+  ) ERC721Rari(_name, _symbol) {
+    _baseUri = _uri;
+    _tokenUriResolver = _tokenUriResolverAddress;
+    _contractUri = _contractMetadataUri;
+  }
+
+  //*********************************************************************//
+  // ---------------------- external transactions ---------------------- //
+  //*********************************************************************//
+
+  /**
+    @notice
+    Owner-only function to set a contract metadata uri to contain opensea-style metadata.
+
+    @param _contractMetadataUri New metadata uri.
+  */
+  function setContractUri(string calldata _contractMetadataUri) external override onlyOwner {
+    _contractUri = _contractMetadataUri;
+  }
+
+  /**
+    @notice
+    Owner-only function to set a new token base uri.
+
+    @param _uri New base uri.
+  */
+  function setTokenUri(string calldata _uri) external override onlyOwner {
+    _baseUri = _uri;
+  }
+
+  /**
+    @notice
+    Owner-only function to set a token uri resolver. If set to address(0), value of baseUri will be used instead.
+
+    @param _tokenUriResolverAddress New uri resolver contract.
+  */
+  function setTokenUriResolver(IJBToken721UriResolver _tokenUriResolverAddress)
+    external
+    override
+    onlyOwner
+  {
+    _tokenUriResolver = _tokenUriResolverAddress;
+  }
+
+  /**
+    @notice
+    Mints the next NFT id.
+
+    @dev
+    Only the owner of this contract can mint more of it.
+
+    ignored: _projectId The ID of the project to which the token belongs. This is ignored.
+    @param _account The account to mint the tokens for.
+  */
+  function mint(uint256, address _account) external override onlyOwner returns (uint256) {
+    uint256 tokenId = _nextTokenId;
```

Comment:
quick trick to avoid natspec/solc complaining of unused var, while keeping to prototype (for reusability):

```
    @param _projectId The ID of the project to which the token belongs. This is ignored.
    @param _account The account to mint the tokens for.
  */
  function mint(uint256 _projectId, address _account) external override onlyOwner returns (uint256) {
      _projectId; // Will not change the bytecode, but solc and natspec happiness are now achieved
      uint256 tokenId = _nextTokenId;
```

Reply:
cool, was wondering why this was done